src/oop/object_interactions/medium/q4/MainTest.java

```diff
@@ -0,0 +1,220 @@
+        assertTrue(testClass.hasModifier(transfusionFromMethodName, new Class<?>[]{Snail.class}, "public"), String.format(incorrectModifierMessage, transfusionFromMethodName, className));
+        assertTrue(testClass.hasReturnType(transfusionFromMethodName, new Class<?>[]{Snail.class}, Void.TYPE), String.format(incorrectReturnType, transfusionFromMethodName, className));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependent1")
+    public void correctTransfusionFromMethod(char type, double volume, char donType, double donVolume) throws Throwable {
+        Object[][] arguments = {
+                {type, char.class},
+                {volume, double.class}
+        };
+        Object[][] donArguments = {
+                {donType, char.class},
+                {donVolume, double.class}
+        };
+        Object donor = testClass.createInstance(donArguments);
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(transfusionFromMethodName, new Class<?>[]{Snail.class}), String.format(incorrectMethodDefinition, transfusionFromMethodName, className));
+        assertTrue(testClass.hasModifier(transfusionFromMethodName, new Class<?>[]{Snail.class}, "public"), String.format(incorrectModifierMessage, transfusionFromMethodName, className));
+        assertTrue(testClass.hasReturnType(transfusionFromMethodName, new Class<?>[]{Snail.class}, Void.TYPE), String.format(incorrectReturnType, transfusionFromMethodName, className));
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {donor, Snail.class}
+        };
+        if (type != donType) {
+            testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Blood type incompatible"),
+                    new Optional(new StringLiteral(" "))
+            });
+        } else if (donVolume <= minBlood) {
+            testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Checking if donor has enough blood for a transfusion\\.\\.\\."),
+                    new Optional(new StringLiteral(" ")),
+                    new NewLine(),
+                    new StringLiteral("Donor does not have enough blood for a transfusion"),
+                    new Optional(new StringLiteral(" "))
+            });
```

Comment:
For the ones where no change should occur, you should also test that the blood volume doesn't change.